### PR TITLE
Add openmodelica header directory #1085

### PR DESCRIPTION
**Pull request by jbwahl**
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/feelpp/feelpp/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes?
- [x] Have you successfully run the Feel++ testsuite with your changes locally?

-----

### CODE REVIEW

**Comment by prudhomm** [comment]
Could you please add some tests in the testsuite to use OMC ?

File: toolboxes/feel/feelmodels/hdg/poroelastic.hpp [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+inline
+po::options_description
+makeMixedPoissonElasticityOptions( std::string prefix = "mixedpoissonelasticity" )
+{
+	po::options_description mpOptions( "Mixed Poisson Elasticity HDG options");
+	mpOptions.add ( makeMixedPoissonOptions("mixedpoisson") );
+	mpOptions.add ( makeMixedElasticityOptions("mixedelasticity") );
+
+	return mpOptions;
+}
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/hdg/poroelastic.hpp]: function should go in a .cpp file. have just the function interface there

File: toolboxes/feel/feelmodels/hdg/poroelastic.hpp [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+
+inline po::options_description
+makeMixedPoissonElasticityLibOptions( std::string prefix = "mixedpoissonelasticity" )
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/hdg/poroelastic.hpp]: same as above

File: toolboxes/feel/feelmodels/hdg/poroelastic.hpp [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+    using vector_type = backend_type::vector_type;
+    using vector_ptrtype = backend_type::vector_ptrtype;
+
+    using Vh_tEL =  Pdhms_type<mesh_type,Order>;
+	using Wh_tEL =  Pdhv_type<mesh_type,Order>;
+	using Vh_tPOI = Pdhv_type<mesh_type,Order>;
+	using Wh_tPOI = Pdh_type<mesh_type,Order>;
+
+	
+	using op_interp_ptrtypeEL = boost::shared_ptr<OperatorInterpolation<Wh_tEL, Pdhv_type<mesh_type,Order>>>;
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/hdg/poroelastic.hpp]: indentation please ;)

File: toolboxes/feel/feelmodels/hdg/poroelastic.hpp [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+    typedef Exporter<mesh_type,1> exporter_type;
+    typedef boost::shared_ptr <exporter_type> exporter_ptrtype;
+    
+
+ 
+ private:
+
+    mesh_ptrtype M_mesh;
+	
+	mp_ptrtype M_PoissonModel;
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/hdg/poroelastic.hpp]: indentation

File: toolboxes/feel/feelmodels/hdg/poroelastic.hpp [hunk 5]
```diff
@@ -0,0 +1,379 @@
+#ifndef _POROELASTIC_HPP
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/hdg/poroelastic.hpp]: please add common Feel++ header here

File: toolboxes/feel/feelmodels/hdg/poroelastic.hpp [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+		this->exportResults (M_PoissonModel->currentTime(), mesh , Idh_el, Idhv_el, Idh_poi, Idhv_poi);
+		M_exporter -> save();
+	}
+
+}; // end class declaration
+
+
+template <int Dim, int Order,int G_Order, int E_Order>
+void
+MixedPoissonElasticity<Dim,Order,G_Order,E_Order>::assembleF_Poisson()
```
Comment (prudhomm) [inline toolboxes/feel/feelmodels/hdg/poroelastic.hpp]: so _ in function names as per coding rules

File: toolboxes/hdg/poroelastic_main.cpp [hunk 7]
```diff
@@ -0,0 +1,118 @@
+#include "../feel/feelmodels/hdg/poroelastic.hpp"
```
Comment (prudhomm) [inline toolboxes/hdg/poroelastic_main.cpp]: add standard header
no ../ in include
use <feel/feelmodels/hdg/poroelastic.hpp>

**Comment by jbwahl** [comment]
Implemented in this pull request : OpenModelica wrapper in Feelpp
The base class of any model is virtual and implemented in `feel/feelom/ommodel.hpp`
During the cmake installation of Feelpp, the declared OM models are generated as well as the specific header and library needed to wrap the model. 
The library is then generated using the proper make command.

Usage : 
- Considering a OM model described in some `.mo` or `.mos` files, in the `CMakelists.txt` of my application, you add the macros 
```
om_add_model( <NAME> SRCS <FILES> LIB <LIBNAME> )
```
with 
<NAME> : the name of the mode. IMPORTANT : it has to be the name of the main class/model in your OM model.
<FILES> : the `.mo` or `.mos` files which describe the OM model. IMPORTANT the main file has to be the first one declared. 
<LIBNAME> the name of the library to be generated. This library has then to be link to your application.

Example of declaration in `testsuite/feelom/CMakeLists.txt' : 
```
om_add_model( test_om SRCS test_om.mo LIB TEST_MO_LIB )
feelpp_add_application( test_om SRCS test_om.cpp LINK_LIBRARIES ${TEST_MO_LIB} )
```

To use the model, you have to add the header `"feel<NAME>.hpp" ' where <NAME> is the name declared in your `CMakeLists.txt`. 
You can then instantiate in your code the wrapper class and run the model with
```
Feel::Feel<NAME> my_om_model;
my_om_model.run();
```
The results of this run are written in the working directory, in the file `<NAME>.csv`

More features coming soon : 
- modification of the input variables through the wrapper
- acquisition of the output data through the wrapper.

**Comment by prudhomm** [comment]
@jbwahl there is a conflict in this PR! we cannot merge currently

**Comment by jbwahl** [comment]
@prudhomm right i noticed that.
I fixed it and i'm working on OM model compilation in cmake. I push it when i'm done.

**Comment by jbwahl** [comment]
New features in the branch: 
- compile an executable from openmodelica model:
```
add_omc( <name> SRCS <srcs_files> )
```
will add the cmake target `feelpp_add_omc_<name>` which compile an executable from the <srcs_files>. The output is in `${CMAKE_CURRENT_BINARY_DIR}/<name>/`. 

WARNING: The first source file has to contain the main class of the model. Also, the <name> has to be the name of this main class!

- create an FMU file from an OpenModelica model:
```
add_fmu( <name> CLASS <class> SRCS <srcs_files> )
```
will add the cmake target `feelpp_add_fmu_<name>` which create an FMU file from OpenModelica `<srcs_files>`. Here the <name> can be set independently of the class name which specified in the <class> variable. The source files have to be ordered to put the main source file in the first position. 

The FMU file can now be loaded in Feelpp with limited features for now. See the test case in `testsuite/feelfmi/'  for an example.

@prudhomm @vincentchabannes I did not manage to do the "install" part of the macros 'add_fmu' and 'add_omc'. 
@prudhomm aside from this last point, I think we are good for a merge.

**Comment by prudhomm** [comment]
We first need to check with buildkite that everything builds at least and also check the testsuite.
could you please rename the macro `feelpp_add_omc` and feelpp_add_fmu`

**Comment by jbwahl** [comment]
@prudhomm should be ok now

**Pull request merged**